src/MonitorProfiler/ProbeInstrumentation/AssemblyProbePrep.cpp

```diff
@@ -0,0 +1,489 @@
+
+    const ULONG expectedLength = (ULONG)m_resolvedCorLibName.length();
+    unique_ptr<WCHAR[]> assemblyName(new (nothrow) WCHAR[expectedLength]);
+    IfNullRet(assemblyName);
+
+    while ((hr = pMetadataAssemblyImport->EnumAssemblyRefs(&corEnum, mdRefs, ENUM_BUFFER_SIZE, &count)) == S_OK)
+    {
+        for (ULONG i = 0; i < count; i++)
+        {
+            mdAssemblyRef curRef = mdRefs[i];
+
+            ULONG nameLength = 0;
+            hr = pMetadataAssemblyImport->GetAssemblyRefProps(
+                curRef,
+                nullptr,
+                nullptr,
+                assemblyName.get(),
+                expectedLength,
+                &nameLength,
+                nullptr,
+                nullptr,
+                nullptr,
+                nullptr);
+
+            if (hr == CLDB_S_TRUNCATION)
+            {
+                // Current assembly's name is longer than corlib's
+                continue;
+            }
+            else if (hr != S_OK)
+            {
+                pMetadataAssemblyImport->CloseEnum(corEnum);
+                return hr;
+            }
+
+            if (nameLength != expectedLength)
+            {
+                continue;
+            }
+
+            tstring assemblyNameStr = tstring(assemblyName.get());
+            if (assemblyNameStr == m_resolvedCorLibName) {
+                pMetadataAssemblyImport->CloseEnum(corEnum);
+                corlibAssemblyRef = curRef;
+                return S_OK;
+            }
+        }
+    }
+
+    if (corEnum)
+    {
+        pMetadataAssemblyImport->CloseEnum(corEnum);
+    }
+
+    IfFailRet(EmitCorLibAssemblyRef(pMetadataEmit, corlibAssemblyRef));
+
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::EmitCorLibAssemblyRef(IMetaDataEmit* pMetadataEmit, mdAssemblyRef& corlibAssemblyRef)
+{
+    IfNullRet(pMetadataEmit);
+
+    HRESULT hr;
+    corlibAssemblyRef = mdAssemblyRefNil;
+
+    ComPtr<IMetaDataAssemblyEmit> pMetadataAssemblyEmit;
+    IfFailRet(pMetadataEmit->QueryInterface(IID_IMetaDataAssemblyEmit, reinterpret_cast<void **>(&pMetadataAssemblyEmit)));
+
+    BYTE publicKeyToken[] = { 0x7c, 0xec, 0x85, 0xd7, 0xbe, 0xa7, 0x79, 0x8e };
+    ASSEMBLYMETADATA corLibMetadata = {};
+    corLibMetadata.usMajorVersion = 4;
+
+    mdAssemblyRef newAssemblyRef;
+    IfFailRet(pMetadataAssemblyEmit->DefineAssemblyRef(
+        publicKeyToken,
+        sizeof(publicKeyToken),
+        m_resolvedCorLibName.c_str(),
+        &corLibMetadata,
+        nullptr,
+        0,
+        afContentType_Default,
+        &newAssemblyRef));
+
+    corlibAssemblyRef = newAssemblyRef;
+
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::HydrateResolvedCorLib()
+{
+    if (m_resolvedCorLibId != 0)
+    {
+        return S_OK;
+    }
+
+    HRESULT hr;
+    ModuleID corLibId = 0;
+    ComPtr<ICorProfilerModuleEnum> pEnum;
+    ModuleID curModuleId;
+    mdTypeDef sysObjectTypeDef = mdTypeDefNil;
+
+    IfFailRet(m_pCorProfilerInfo->EnumModules(&pEnum));
```

Comment:
You might have to validate if this is still true, but it used to be the case that the corelib would always load first in the ModuleLoaded callback. Rather than attempting to resolve corlib by enumerating all the modules, you can simply propagate it from the load event.